Hi Pratap Roy! Thanks for working on this project for our team.

As a data-driven finance company, we rely on data to make informed decisions that drive our day-to-day operations. However, we have noticed that some of our operational team members are relying on their own assumptions and interpretations of the data, leading to inaccurate decision-making. While some assumptions may prove to be accurate, we need to ensure that all our decisions are based on sound statistical analysis. To address this challenge, we need to evaluate a series of assumptions related to our operations based on factual data.

As a data scientist in my team, you are assigned with evaluating these assumptions using hypothesis testing and statistical tests in Python and Jupyter Notebook and conclude whether a certain assumption is right or wrong, based on the given data. You will be provided with a dataset containing relevant variables for each assumption and will be expected to use the appropriate statistical methods to analyze the data and draw conclusions. Your goal will be to evaluate each assumption using appropriate statistical methods, draw conclusions based on the data, and present your findings to the management team. By completing this project, you will contribute to our goal of data-driven decision making.

### Load the data and check the data quality

In [14]:
import pandas as pd
 
df = pd.read_csv('loan-dataset-1.csv')
 
df.head()

loan_id  principal_loan_amount request_date  \
0  70280c1b-4339-4949-ba69-ad2a6cb6b782                   8880   2023-03-06   
1  8a550559-9749-491b-b6aa-7188bc0a488f                  13720   2023-05-10   
2  5644d27e-e05e-4a40-bef3-64187892338f                   2090   2023-02-16   
3  9d8d17ac-8af9-4936-ba5d-8d5ed5e0cfab                   6820   2023-02-26   
4  4586ad7a-2231-419b-b8e3-ba03b929b56e                  13770   2023-04-13   

   interest_rate loan_channel is_approved is_default  customer_monthly_income  \
0            9.0          WEB        True        NaN                     6100   
1           10.0   MOBILE_APP        True        NaN                     5150   
2            7.5        AGENT        True        NaN                     6030   
3            8.5   MOBILE_APP         NaN        NaN                     2150   
4           10.0    AFFILIATE         NaN        NaN                     5810   

   customer_age customer_gender  customer_credit_score  customer_zip_code  
0            30            Male                      3              10007  
1            39          Female                      3              10004  
2            43          Female                      3              10003  
3            54            Male                      0              10004  
4            28          Female                      2              10007

In [15]:
df.shape

(15163, 12)

In [16]:
df.duplicated().any()

True

In [29]:
df.isnull().sum()

loan_id                    0
principal_loan_amount      0
request_date               0
interest_rate              0
loan_channel               0
is_approved                0
is_default                 0
customer_monthly_income    0
customer_age               0
customer_gender            0
customer_credit_score      0
customer_zip_code          0
dtype: int64

In [18]:
 print(df.isna().any(axis=0)[df.isna().any()], '\n')

is_approved    True
is_default     True
dtype: bool 



In [19]:
print('principal_loan_amount:', df['principal_loan_amount'].min(), 'to', df['principal_loan_amount'].max())
print('request_date:', df['request_date'].min(), 'to', df['request_date'].max())
print('interest_rate:', df['interest_rate'].min(), 'to', df['interest_rate'].max())
print('loan_channel:', df['loan_channel'].unique())
print('is_approved:', df['is_approved'].unique())
print('is_default:', df['is_default'].unique())
print('customer_monthly_income:', df['customer_monthly_income'].min(), 'to', df['customer_monthly_income'].max())
print('customer_age:', df['customer_age'].min(), 'to', df['customer_age'].max())
print('customer_gender:', df['customer_gender'].unique())

principal_loan_amount: 1000 to 15000
request_date: 2023-01-01 to 2023-06-29
interest_rate: 7.5 to 10.0
loan_channel: ['WEB' 'MOBILE_APP' 'AGENT' 'AFFILIATE' 'DIRECT_SELLING'
 'website_revamped' 'ANDROID_V2' 'ANDROID_V3' 'apple_new_v3']
is_approved: [True nan]
is_default: [nan True]
customer_monthly_income: 2000 to 6500
customer_age: 22 to 60
customer_gender: ['Male' 'Female' 'M' 'F']


### Data Cleaning

In [20]:
df[df.duplicated(keep = False)]

loan_id  principal_loan_amount  \
166    5fb226e9-d3d7-4f63-9c6a-6b842584b860                  10750   
167    5fb226e9-d3d7-4f63-9c6a-6b842584b860                  10750   
196    e5ce6885-92d7-4e57-9f1e-8e3bf31c935a                   6790   
197    e5ce6885-92d7-4e57-9f1e-8e3bf31c935a                   6790   
330    ca4a6603-0bd2-4ee9-8f60-54061d75b3a1                  11420   
...                                     ...                    ...   
14897  84c0e3dc-071e-4d44-bee1-c8c7cc713577                   7520   
15000  6be039f6-562d-4149-937b-470e6d7549a5                  10920   
15001  6be039f6-562d-4149-937b-470e6d7549a5                  10920   
15077  688ceb23-8cba-4c37-bc3d-f1789a36ab93                  10160   
15078  688ceb23-8cba-4c37-bc3d-f1789a36ab93                  10160   

      request_date  interest_rate loan_channel is_approved is_default  \
166     2023-04-16            9.5   MOBILE_APP        True        NaN   
167     2023-04-16            9.5   MOBILE_APP        True        NaN   
196     2023-04-15            8.5    AFFILIATE        True        NaN   
197     2023-04-15            8.5    AFFILIATE        True        NaN   
330     2023-04-16            9.5          WEB        True       True   
...            ...            ...          ...         ...        ...   
14897   2023-04-15            9.0    AFFILIATE        True        NaN   
15000   2023-04-15            9.5    AFFILIATE         NaN        NaN   
15001   2023-04-15            9.5    AFFILIATE         NaN        NaN   
15077   2023-04-15            9.5        AGENT        True        NaN   
15078   2023-04-15            9.5        AGENT        True        NaN   

       customer_monthly_income  customer_age customer_gender  \
166                       2480            44            Male   
167                       2480            44            Male   
196                       4520            42          Female   
197                       4520            42          Female   
330                       4780            52          Female   
...                        ...           ...             ...   
14897                     4650            54            Male   
15000                     2300            26            Male   
15001                     2300            26            Male   
15077                     5930            43            Male   
15078                     5930            43            Male   

       customer_credit_score  customer_zip_code  
166                        3              10004  
167                        3              10004  
196                        3              10006  
197                        3              10006  
330                        3              10003  
...                      ...                ...  
14897                      3              10005  
15000                      3              10004  
15001                      3              10004  
15077                      3              10006  
15078                      3              10006  

[326 rows x 12 columns]

In [21]:
df = df.drop_duplicates()
df.shape

(15000, 12)

In [22]:
df['is_approved'].fillna(False, inplace= True)
df['is_default'].fillna(False, inplace = True)

In [23]:
mapping_loan_chanel = {
    'WEBSITE_REVAMPED': 'WEB',
    'ANDROID_V2': 'MOBILE_APP',
    'ANDROID_V3': 'MOBILE_APP',
    'APPLE_NEW_V3': 'MOBILE_APP'
}

df['loan_channel'] = df['loan_channel'].str.upper().replace(mapping_loan_chanel)
df['loan_channel'].unique()

array(['WEB', 'MOBILE_APP', 'AGENT', 'AFFILIATE', 'DIRECT_SELLING'],
      dtype=object)

In [33]:
df['customer_gender'] = df['customer_gender'].str.upper().replace(['MALE','FEMALE'],['M','F'])
df['customer_gender'].unique()

array(['M', 'F'], dtype=object)

In [35]:
df.to_csv('loan-dataset-1-clean.csv', index = False)

### Evaluate the impact of Interest Rates on Loan Default Rates

Null hypothesis (from the operations team) : the interest rate has no impact on the loan default rate

Alternative hypothesis : interest rate does have an impact on the loan default rate.

Use the two-tailed t-test to evaluate the impact of interest rates on loan default rates. Split the data into two groups based on the interest rate levels, then compare the mean default rates between the two groups to analyze the hypothesis.



In [37]:
df['interest_rate'].describe()

count    15000.000000
mean         8.853000
std          0.814781
min          7.500000
25%          8.000000
50%          9.000000
75%          9.500000
max         10.000000
Name: interest_rate, dtype: float64

In [38]:
import scipy.stats as stats

In [39]:
median_interest = df['interest_rate'].median()
median_interest

9.0

In [40]:
high_int = df[df['interest_rate'] > median_interest]
low_int = df[df['interest_rate'] <= median_interest]

In [51]:
# The hypothesis is about the relationship between interest rate and default rate. 
# Calculate the default rate's average (mean) for both interest rate groups.
high_default_rate = high_int['is_default'].mean()
low_default_rate = low_int['is_default'].mean()

Use $stats.ttest_ind()$ function from $scipy.stats$ library to compare the default rate of loans with high interest rates versus the default rate of loans with low interest rates, based on $is_default$ column of the $high_interest$ and $low_interest$ dataframes.

Assign output from $ttest_ind()$ function into:

- t_stat (t-statistic to measure the difference between the means of the two groups divided by the standard error of the difference)

- p_val (p-value to indicate how strong the evidence is against the null hypothesis)

In [54]:
t_stat, p_val = stats.ttest_ind(high_int['is_default'], low_int['is_default'])
t_stat, p_val

(-1.1830953524322996, 0.23679011271197217)

In [55]:
print('High interest rate default rate: {:.2%}'.format(high_default_rate))
print('Low interest rate default rate: {:.2%}'.format(low_default_rate))
print('Two-tailed t-test: t-statistic = {:.2f}, p-value = {:.2f}'.format(t_stat, p_val))

High interest rate default rate: 3.37%
Low interest rate default rate: 3.74%
Two-tailed t-test: t-statistic = -1.18, p-value = 0.24


The t-statistic is -1.18 meaning that the average (mean) default rate for loans with high interest rates is slightly lower than the mean default rate for loans with low interest rates.

However, the p-value is 0.24, which is higher than the commonly used threshold of 0.05 for statistical significance.

This means that the difference in default rates between the two groups may not be statistically significant. In other words, the interest rate has no significant impact on loan default rate.

> > > Hence the null hypothesis (the interest rate has no impact on the loan default rate) is $True$

### Evaluate the relationship between Income Level and Loan Approval Rates

Null hypothesis (from the operations team) : customer income level has no impact on the loan approval rate

Alternative hypothesis : customer income level impacts the loan approval rate

Use logistic regression analysis to evaluate the impact of customer income level on loan approval rates. The customer income levels are separated into these groups:

- less than 2500

- 2500 to 5000

- 5000 to 7500

- more than 7500

In [58]:
import numpy as np
import statsmodels.api as sm

In [59]:
income_groups = pd.cut(df['customer_monthly_income'], bins=[0, 2500, 5000, 7500, np.inf])
approval_rate = df.groupby(income_groups)['is_approved'].mean()
approval_rate

customer_monthly_income
(0.0, 2500.0]       0.781103
(2500.0, 5000.0]    0.776105
(5000.0, 7500.0]    0.775494
(7500.0, inf]            NaN
Name: is_approved, dtype: float64

In [62]:
df[df['customer_monthly_income'] >= 7500]

Empty DataFrame
Columns: [loan_id, principal_loan_amount, request_date, interest_rate, loan_channel, is_approved, is_default, customer_monthly_income, customer_age, customer_gender, customer_credit_score, customer_zip_code]
Index: []

- Use is_approved for dependent variable (variable to explain)

- Use customer_monthly_income for independent variable (variable to explain the variation in the dependent variable)

- Use Logit function to fit a logistic regression model to the data.

In [63]:
X = sm.add_constant(df['customer_monthly_income'])
y = df['is_approved']
model = sm.Logit(y, X).fit()
print(model.summary())

Optimization terminated successfully.
         Current function value: 0.531339
         Iterations 5
                           Logit Regression Results                           
Dep. Variable:            is_approved   No. Observations:                15000
Model:                          Logit   Df Residuals:                    14998
Method:                           MLE   Df Model:                            1
Date:                Fri, 19 May 2023   Pseudo R-squ.:               8.514e-06
Time:                        04:16:33   Log-Likelihood:                -7970.1
converged:                       True   LL-Null:                       -7970.2
Covariance Type:            nonrobust   LLR p-value:                    0.7126
                              coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------
const                       1.2688      0.067     18.948      0.000       1.138   

The coefficient for the customer_monthly_income is -5.539e-06, which means that one-unit increase in the customer_monthly_income results in a -5.539e-06 unit decrease in the log odds of loan approval (is_approved).

The p-value for this coefficient is 0.713, which is greater than the significance level of 0.05. Therefore, we fail to reject the null hypothesis that there is no linear relationship between income level and loan approval rates.

### Find variables that affect loan default rate

Since interest rate does not impact the default rate, you need to find out possible variable(s) that affect loan default rate.

The risk team suggested the following variable possibilities:

- customer_monthly_income

- customer_zip_code

You need to use regression test against data to find out whether the risk team's suggestion is correct.

In [64]:
X = sm.add_constant(df[['customer_monthly_income','customer_zip_code']])
y = df['is_default']
model = sm.Logit(y, X).fit()
print(model.summary())

         Current function value: 0.045512
         Iterations: 35
                           Logit Regression Results                           
Dep. Variable:             is_default   No. Observations:                15000
Model:                          Logit   Df Residuals:                    14997
Method:                           MLE   Df Model:                            2
Date:                Fri, 19 May 2023   Pseudo R-squ.:                  0.7068
Time:                        04:23:12   Log-Likelihood:                -682.68
converged:                      False   LL-Null:                       -2328.5
Covariance Type:            nonrobust   LLR p-value:                     0.000
                              coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------
const                    3.662e+04   1640.488     22.324      0.000    3.34e+04    3.98e+04
customer_monthly_income -1

/Users/prataprc94/opt/anaconda3/lib/python3.8/site-packages/statsmodels/base/model.py:604: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


- p-value for customer_monthly_income is greater than the significance level of 0.05, which means customer_monthly_income does not affect default rate

- p-value for customer_zip_code is less than the significance level of 0.05, which means customer_zip_code significantly affect default rate

- However, warning message Maximum Likelihood optimization failed to converge suggests that the model may not have converged thus the results should be interpreted with caution

- The warning Possibly complete quasi-separation indicates that there may be perfect prediction or separation of the data points (in this case customer_zip_code perfectly affects default rate), so the results should be interpreted with caution

### Evaluate loan channel against Default Rate

The operations team came up with these additional hypotheses:

Null hypothesis: loan channel has no impact on the loan approval rate

Alternative hypothesis: loan channel impacts the loan approval rate

Use ANOVA test to determine if there is a significant difference between loan default rates across different loan channels. If the test indicates that there is a significant difference, you can use post-hoc tests to determine which specific loan channels have different loan default rates.

In [69]:
import statsmodels.formula.api as smf

In [70]:
df['is_default'] = df['is_default'].astype(int)
df['is_default'].unique()

array([0, 1])

Perform the ANOVA test. Specify is_default as the dependent variable, and loan_channel is the categorical independent variable.

Calculate ANOVA table for the regression model using the anova_lm function from statsmodels.stats.anova. Use Type II ANOVA to determine the significance of each independent variable in the presence of other independent variables.

In [71]:
model = smf.ols("is_default ~ C(loan_channel)", data = df)
results = model.fit()
aov_table = sm.stats.anova_lm(results, typ = 2)
print(aov_table)

                     sum_sq       df           F         PR(>F)
C(loan_channel)   29.296879      4.0  223.138301  2.286606e-186
Residual         492.191054  14995.0         NaN            NaN


> The p-value is less than significance level of 0.05, means there is a statistically significant difference between the groups of loan_channel.

Perform post-hoc tests to determine which loan channel has different impact on default rate. Use Tukey's Honestly Significant Difference (HSD) test on the data. Use is_default as the response variable and loan_channel column as the predictor variable.

In [72]:
mc = sm.stats.multicomp.MultiComparison(df['is_default'], df['loan_channel'])
mc_results = mc.tukeyhsd()
print(mc_results.summary())

       Multiple Comparison of Means - Tukey HSD, FWER=0.05        
    group1         group2     meandiff p-adj  lower  upper  reject
------------------------------------------------------------------
     AFFILIATE          AGENT  -0.0034   0.9 -0.0161 0.0093  False
     AFFILIATE DIRECT_SELLING   0.0004   0.9 -0.0125 0.0133  False
     AFFILIATE     MOBILE_APP  -0.0036   0.9 -0.0165 0.0092  False
     AFFILIATE            WEB   0.1059 0.001  0.0934 0.1185   True
         AGENT DIRECT_SELLING   0.0038   0.9 -0.0091 0.0167  False
         AGENT     MOBILE_APP  -0.0002   0.9 -0.0131 0.0126  False
         AGENT            WEB   0.1093 0.001  0.0968 0.1218   True
DIRECT_SELLING     MOBILE_APP   -0.004   0.9  -0.017  0.009  False
DIRECT_SELLING            WEB   0.1055 0.001  0.0928 0.1182   True
    MOBILE_APP            WEB   0.1095 0.001  0.0969 0.1222   True
------------------------------------------------------------------


Examine the post-hoc test. If the reject column is true for a pair of loan channels, then you can reject the null hypothesis and conclude that there is a significant difference between the default rates of the loan channels.

In the post-hoc table, the reject column is true for the pairs of loan channels (AFFILIATE, WEB), (AGENT, WEB), (DIRECT_SELLING, WEB), and (MOBILE_APP, WEB). This suggests that the default rate for loans obtained through the WEB channel is significantly different from the default rates for loans obtained through the other channels.

#### Hence we can reject the null hypothesis, since WEB channel contributes to loan default rate